### ISSUE TITLE
Integrate AppImage generation tools into IDEs and build tools

### ISSUE THREAD

**Issue by probonopd** [probonopd/AppImageKit#160]
It would be nice if we could make it super easy for developers to produce AppImages. For this, integration into IDEs like Qt Creator and build tools like electron-builder might be helpful. Let's collect tools for which this might be useful here (please feel free to add!).

I am happy to help any upstream IDE or build tool wishing to integrate AppImage generation.

* [Qt Creator](https://www.qt.io/ide/)
* [electron-builder](https://github.com/electron-userland/electron-builder): [GitHub Issue](https://github.com/electron-userland/electron-builder/issues/504)

**Comment by chrisnharvey** [comment]
Yes! I love this idea.

How about GNOME Builder?

**Comment by probonopd** [comment]
Added

**Comment by shoogle** [comment]
While IDE integration would be very welcome, I think lower-level integration via build systems like Make and CMake would be more useful in practice. Everyone has their personal favourite IDE (or none if they prefer text editors and the command line), so the AppImage build process needs to be IDE-agnostic. The best (most portable) solution would be for AppImageKit to provide a series of generic build scripts that could be called via a special Makefile target. ([Here's the target](https://github.com/musescore/MuseScore/blob/master/Makefile#L132-L146) I made for MuseScore.) IDE integration would just take the form of a special "Build AppImage" button that would run the Makefile target.

**Comment by probonopd** [comment]
True @shoogle. Can you help? In the ideal world we would have some sort of cross-compiler system that would use recent versions of CMake and the like to build for mature systems like CentOS 6, debian oldstable etc. to ensure maximum binary compatibility. Right now this is way over my head though.

**Comment by cgalvarez** [comment]
Hi @probonopd. I've successfully generated the `.AppImage` from a Meteor project and its related `.zsync`, but I would like to make it self-updateable.

From your [AppImageUpdate README](https://github.com/probonopd/AppImageKit/blob/master/AppImageUpdate.AppDir/README.md#example-how-to-make-your-app-self-updateable):

> You could put appimageupdate (and its dependency zsync_curl) inside usr/bin of the AppImage, and call appimageupdate from within your app and even have a little GUI around it; or reimplement it in the language of your app.

I've already downloaded `appimageupdate` and `zsync_curl` to `usr/bin` inside the AppImage, and added that folder into `process.env.PATH` (Meteor runs over Node.js), but I have no success when executing the command `/bin/sh -c appimageupdate meteor-app.AppImage set "zsync|http://localhost:3000/download/linux/meteor-app.AppImage.zsync"`. I get the error `meteor-app.AppImage: 1: meteor-app.AppImage: appimageupdate: not found`. I'm running that command inside the Electron client, which is wrapped as an AppImage.

Should I run it that command just after building the .AppImage over it? Should `meteor-app.AppImage` in that command be the location of the AppImage executable?

**Comment by probonopd** [comment]
Try `/place/where/you/downloaded/appimageupdate /path/to/your/AppAppImage`.

**Comment by cgalvarez** [comment]
Thanks for your help, @probonopd. Took it just after writing the comment.

If I bundle `appimageupdate` and `zsync_curl` with the AppImage, then the `path/to/your/AppAppImage/` is the location where the downloaded executable is placed. The app contents are mounted in a temp folder at `/tmp`, but what I am wondering now is right how to get that path of the downloaded AppImage... because the user can place it wherever s/he wants.

**Comment by probonopd** [comment]
@cgalvarez can you update the AppImage by using [AppImageUpdate](https://bintray.com/probono/AppImages/AppImageUpdate/206.g78dc97c/view#files) outside of your AppImage?

I understand that you want to make the AppImage update "itself" without the need for an external AppImageUpdate.AppImage. Correct? In that case, you currently need to bundle curl (at least until someone closes https://github.com/probonopd/zsync-curl/issues/9 - should not be hard), zsync_curl, and appimageupdate. Possibly you might want to bundle appimageupdategui as well, to have a GUI updater.

Then you need to point the updater to the AppImage that is currently being used ("itself"). You can use the environment variable `$APPIMAGE` for that.

**Comment by cgalvarez** [comment]
@probonopd Yes, I'tve tested updating my AppImage from outside (from the command line in local) successfully.

You're right. I want to make the AppImage update itself without any intervention, so it should point to itself (what I was poorly saying *wherever the user wants to download it, sorry). Didn't know about the env var `$APPIMAGE`. Thanks!!

I was getting the path through the command `ps aux | grep 'app.AppImage$'` when you wrote your answer.

Tomorrow I will git it another try. Thank you very much for your explanations and your promptness.

**Comment by hbirchtree** [comment]
Hello,
I was contacted by probonopd about CMake integration. I did the approach of packaging each executable CMake target by themselves with necessary libraries and data files.
Some of the assumptions made could be wrong (such as classifying data only into the categories of executables, libraries and data), but this is what I think works.
Here is the latest version (as of the time of this post) I worked on: [Link](https://github.com/hbirchtree/coffeecutie/blob/f83576fe3bb5a156db576b52b4d8146c274ef56a/cmake/Modules/LinuxAppImageBuild.cmake)

**Comment by probonopd** [comment]
Hello @hbirchtree welcome and thanks for your CMake script. Could you give some instructions for people not familiar with CMake on how to use it, e.g., to package a simple "hello world" app like https://gist.github.com/fracek/3323924?

**Comment by hbirchtree** [comment]
@probonopd here you go:
https://gist.github.com/hbirchtree/afada1b4886a79abac3e2771773059e3
You will have to specify paths to the AppRun file and AppImageAssistant manually (APPIMAGE_APPRUN_PROGRAM and APPIMAGE_ASSISTANT_PROGRAM). I assume these will have reliable paths in the future?

**Comment by probonopd** [comment]
Thank you very much @hbirchtree, your help is really appreciated. I have commented on some issues I identified on https://gist.github.com/hbirchtree/afada1b4886a79abac3e2771773059e3

**Comment by probonopd** [comment]
We should have a deep look at
http://code.qt.io/cgit/qt/qttools.git/tree/src/macdeployqt/
and see if the same can be done to generate an AppImage.

**Comment by probonopd** [comment]
Started to work on [linuxdeployqt](https://github.com/probonopd/linuxdeployqt). Please test, review, and send pull requests.

**Comment by probonopd** [comment]
[coffeecutie-automation-tools](https://github.com/hbirchtree/coffeecutie-automation-tools/blob/master/builders/README.md)

> Builds for Linux against Steam Runtime
Mostly independent of underlying platform
Based on Steam Runtime chroot setup (but better)
Not fully automated yet
__Generates standard binaries, can also generate AppImage (support incoming)__
Has Ninja generator

**Comment by probonopd** [comment]
CMake's CPack can produce `.app` bundles and `.dmg` disk images for macOS, so we should try to get it produce AppImages as well:

https://cmake.org/Wiki/CMake:CPackPackageGenerators#Bundle_.28OSX_only.29

**Comment by probonopd** [comment]
AppImageAssistant has been replaced by `appimagetool`. Use it instead.

**Comment by probonopd** [comment]
Now that we have CMake experts @TheAssassin @azubieta do you think we could make an AppImage equivalent of https://cmake.org/Wiki/CMake:CPackPackageGenerators#Bundle_.28OSX_only.29?

**Comment by TheAssassin** [comment]
You're clearly mistaking "average CMake skills" for "is able to write a CPack module".

**Comment by nocnokneo** [comment]
I've started the work to add AppImage support to CPack here: https://gitlab.kitware.com/cmake/cmake/merge_requests/2645/

It's definitely not polished yet but it's still usable for early testing. I'll try find time to get it cleaned up and into the next CMake release.

**Comment by TheAssassin** [comment]
Very cool. Will check it out more detailedly soon, but at a first glance, this looks quite good for an alpha version.

You should check out https://github.com/linuxdeploy/linuxdeploy as a (more modern and flexible) alternative to linuxdeployqt. linuxdeploy is less monolithic, it only bundles binary dependencies and a few auxiliary data (like desktop files, icons, soon probably even AppStream).
It features a powerful plugin system, allowing users to add additional functionality, like e.g., [bundling Qt dependencies](https://github.com/linuxdeploy/linuxdeploy-plugin-qt/). AppImage creation itself is [a plugin](https://github.com/linuxdeploy/linuxdeploy-plugin-appimage).
Environment variables are used to configure plugin behavior. You can ship custom AppRuns with a parameter, etc. etc. etc.

I hope that shows why it's a lot more flexible than linuxdeployqt. Oh, and btw, linuxdeploy is way better tested.

**Comment by nocnokneo** [comment]
Thanks for the tip. I wasn't aware of linuxdeploy. I'll check it out when I get a chance. Eventually the AppImage CPack generator could support multiple tools depending on what it can find on the system, but initially it probably makes sense to keep it simple and support just one (linuxdeploy sounds like the best option)

**Comment by probonopd** [comment]
Welcome to AppImage, @nocnokneo. Great work!

**Comment by danger89** [comment]
AppImage within Cpack would be ideal!

**Comment by probonopd** [comment]
Check out 

* https://gitlab.kitware.com/cmake/cmake/-/issues/18782
* https://gitlab.kitware.com/cmake/cmake/-/merge_requests/2645

Apparently @azubieta knows something about CPack (I don't really, CMake is a `$UNKNOWN` `$MYSTERY` to `$MYSELF`).

**Comment by Shatur95** [comment]
Currently can integrated with CMake by using [CPack External Generator](https://cmake.org/cmake/help/latest/cpack_gen/external.html).

You can take a look at [this](https://gist.github.com/slurps-mad-rips/019106e20b297174abf30a8f74eff9b3) gist for an example.

[This](https://github.com/crow-translate/crow-translate/commit/fe226a16971df0d42b62777da19a3aa1ae7117e9) is how I personally integrated it into my Qt project.
